broker/src/bin/bench.rs

```diff
@@ -0,0 +1,175 @@
+use neon_broker::neon_broker_proto::subscribe_safekeeper_info_request::SubscriptionKey;
+use neon_broker::neon_broker_proto::TenantTimelineId as ProtoTenantTimelineId;
+use neon_broker::neon_broker_proto::{SafekeeperTimelineInfo, SubscribeSafekeeperInfoRequest};
+use neon_broker::DEFAULT_LISTEN_ADDR;
+use tokio::time::{self};
+
+use tonic::transport::Channel;
+use tonic::Request;
+
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct Args {
+    /// Number of publishers
+    #[clap(short = 'p', long, value_parser, default_value_t = 1)]
+    num_pubs: u64,
+    /// Number of subscribers
+    #[clap(short = 's', long, value_parser, default_value_t = 1)]
+    num_subs: u64,
+}
+
+async fn progress_reporter(counters: Vec<Arc<AtomicU64>>) {
+    let mut interval = time::interval(Duration::from_millis(1000));
+    let mut c_old = counters.iter().map(|c| c.load(Ordering::Relaxed)).sum();
+    let mut c_min_old = counters
+        .iter()
+        .map(|c| c.load(Ordering::Relaxed))
+        .min()
+        .unwrap_or(0);
+    let mut started_at = None;
+    let mut skipped: u64 = 0;
+    loop {
+        interval.tick().await;
+        // print!(
+        //     "cnts are {:?}",
+        //     counters
+        //         .iter()
+        //         .map(|c| c.load(Ordering::Relaxed))
+        //         .collect::<Vec<_>>()
+        // );
+        let c_new = counters.iter().map(|c| c.load(Ordering::Relaxed)).sum();
+        let c_min_new = counters
+            .iter()
+            .map(|c| c.load(Ordering::Relaxed))
+            .min()
+            .unwrap_or(0);
+        if c_new > 0 && started_at.is_none() {
+            started_at = Some(Instant::now());
+            skipped = c_new;
+        }
+        let avg_rps = started_at.map(|s| {
+            let dur = s.elapsed();
+            let dur_secs = dur.as_secs() as f64 + (dur.subsec_millis() as f64) / 1000.0;
+            let avg_rps = (c_new - skipped) as f64 / dur_secs;
+            (dur, avg_rps)
+        });
+        println!(
+            "sum rps {}, min rps {} total {}, total min {}, duration, avg sum rps {:?}",
+            c_new - c_old,
+            c_min_new - c_min_old,
+            c_new,
+            c_min_new,
+            avg_rps
+        );
+        c_old = c_new;
+        c_min_old = c_min_new;
+    }
+}
+
+fn tli_from_u64(i: u64) -> Vec<u8> {
+    let mut timeline_id = vec![0xFF; 8];
+    timeline_id.extend_from_slice(&i.to_be_bytes());
+    timeline_id
+}
+
+async fn subscribe(client: Option<NeonBrokerClient<Channel>>, counter: Arc<AtomicU64>, i: u64) {
+    let mut client = match client {
+        Some(c) => c,
+        None => NeonBrokerClient::connect(format!("http://{}", DEFAULT_LISTEN_ADDR))
+            .await
+            .unwrap(),
+    };
+
+    // let key = SubscriptionKey::All(());
+    let key = SubscriptionKey::TenantTimelineId(ProtoTenantTimelineId {
+        tenant_id: vec![0xFF; 16],
+        timeline_id: tli_from_u64(i),
```

Comment:
If you would like to see fixed size array here, then no, protobuf doesn't support them at all. Yes, somewhat disappointing, but here I don't think this matters.

Reply:
Even without fixed size arrays vec shouldnt be necessary. It should be possible to populate the value from &[u8] or an iterator. There is even an issue in the library for that (not mentioning it to avoid direct linking to it). Two allocations for each message look bad for me

Reply:
This still looks like a problem to me. Lets ask someone else to get third opinion